ncid.1
.\" %W% %G%
.TH NCID 1
.SH NAME
.B ncid\^
- Network Caller ID Client
.SH SYNOPSIS
.B ncid\^
[options] [arguments]
.P
.B tivocid\^
[options] [arguments]
.P
.I Options:\^
.P
.nf
[--no-gui]
[--all             | -A]
[--call-prog       | -C]
[--delay seconds   | -D seconds]
[--program PROGRAM | -P PROGRAM]
[--raw             | -R]
[--verbose         | -V]
[--ring count      | -r count]
.fi
.P
.I Arguments:\^
.P
.nf
[IP_ADDRESS | HOSTNAME]
[PORT_NUMBER]
.fi
.SH DESCRIPTION
.B NCID
displays Caller ID information sent to it by the Caller ID server.
In the default GUI mode, if the server is configured to send the
CID Call Log, it will load the call log in the text window at
startup.  The call log is updated by \fIncid\fR whenever a call
is received.
.P
.B NCID
runs on Linux and Windows 98, ME, XP, 2000, NT.  In the GUI mode, it
will pop up, and go on top whenever a call comes in.  In addition,
it will display itself on whatever desktop is active in Linux.
If \fIncid\fR is given the --no-gui option, it will either output
to a external program or use the verbose mode to display the CID
information in a terminal window.
.P
If \fIncid\fR is named \fItivocid\fR, it will start in a no-gui mode
and output to a external program called \fIout2osd\fR which displays
the Caller ID on a TV using a TiVo.
.P
If \fIncid\fR is configured to output to a external program, it will
test for it, and if it is not found, or if it is not executable,
it will terminate with a error message.
.P
The configuration file for \fIncid\fR is /etc/ncid/ncid.conf.
.P
Three external scripts are provided, customize as needed in the
/etc/ncid/ncidscript.conf file:
.TP 11
.PD 0
.B ncid-page
Sends a page or cell phone notification, with the CID information,
when a call is received.
Must set $ADDRESS, in the script, to the pager address.
.TP
.B ncid-speak
Speaks the caller name when a call is received, using \fIfestival\fB.
Script can be modified to speak any of the CID information.
.TP
.B ncid-samba
Send a SMB popup message containing the Caller ID information.
.PD
.P
Whenever the server connection is broken, \fIncid\fR will try to reconnect
once a minute.  It provides a visual record of the countdown and number
of tries.
.P
The file menu provides three menu items in the GUI mode:
.TP 11
.PD 0
.B reconnect
provides an immediate reconnect and a reload of the call log file.
This provides a way to synchronize \fIncid\fR with the server call log in
the event of a network problem.
.TP
.B clear
clears the call log
.TP
.B quit
terminates \fIncid\fR.
.PD
.SS "Options"
.PD 0
.TP 30
.B --no-gui
Do not display CID information using a GUI
.br
Default: display CID using a GUI, except if called tivocid
.TP
.B --all | -A
output: date time number name
.br
Default output: name number
.TP
.B --call-prog | -C
output name and number to external program
.br
Default: does not output to external program unless name is tivocid
.TP
.B --delay seconds | -D seconds
delay between server connection attempts
.br
Default: 60 seconds
.TP
.B --prog PROGRAM | -P PROGRAM
the display or helper program to use
.br
Default PATH: /usr/share/ncid
.br
Default PROGRAM: ncid-speak
.TP
.B --raw | -R
display data received from the CID Server
.br
Default: raw data not displayed
.TP
.B --ring count | -r count
call external program on ring count
.br
Default: external program called as soon as CID received
.TP
.B --verbose | -V
display formatted CID information, including CID log file
.br
Default: silent
.PD
.SS "Arguments"
.PD 0
.TP 30
.B IP_ADDRESS | HOSTNAME
set the IP Address
.br
Default: 127.0.0.1 (localhost)
.TP
.B PORT_NUMBER
set the port number
.br
Default: 3333
.PD
.SH FILES
/etc/ncid/*
.br
/usr/share/ncid/*
.SH EXAMPLES
.nf
# run ncid in a terminal window, server on remote host
ncid --no-gui tardis.drwho.home
# output Caller ID to external program
ncid --no-gui --call-prog --program ncid-speak
# output NAME & NUMBER to GUI and external program
ncid --call-prog --program /usr/local/bin/ncid-speak
# output NAME & NUMBER to GUI and all CID info to external program
ncid --all --call-prog --program ncid-page
# output all CID info to external program after third ring
ncid --no-gui --all --call-prog --ring 3 --program ncid-page
# run ncid on a TiVo, server on remote host
tivocid 192.168.0.1
ncid --no-gui --call-prog --program /var/hack/bin/out2osd 192.168.0.1
.fi
.SH SEE ALSO
lcdncid.1, out2osd.1, ncidd.8, ncidd.conf.5 ncidd.alias.5


doc/CHANGES
ncid - Network CallerID

Version 0.5
    - Released

Version 0.5.1
    - Modem port open changed so it would not hang if
      port in use, or if port was not restored after use.
      It will fail sometimes.
    - corrected MODEMLOG value in Makefile
    - DATE field in CID line modified to include year: MMDDYYYY
    - Changed FORMAT and ncidd.8 to reflect new DATE field.

Version 0.6
    - ncidd.c: increased the number of TCP connections from 9 to 15
    - ncidd.c: added routines to send the CID call log file to clients
    - ncidd.c: doModem(): eliminated pointer "ptr"
    - ncidd.c: doModem(): added tcflush()
    - ncidd.c: doModem(): added new error message "No modem found"
      for no modem response
    - ncidd.c: doModem(): put call to initialize modem in a for loop
      to try 5 times, sometimes it failed to get a modem response
      the 1st time
    - ncidd.c doPoll(): added check; if fd == 0, set revents == 0
      sometimes the poll seemed to get stuck on revents for fd = 0
    - ncidd.c: doPoll(): modified "if (addPoll(sd) < 0)" to send
      message to client, and write message to stderr if "verbose"
    - ncidd.c doPoll(): added clear socker errors when client
      disconnects
    - ncidd.c: TIMEOUT 10 -> 50
    - ncidd.c: CONNECTIONS + 3 -> CONNECTIONS + 2
    - ncidd.c: cleanup(): added tcflush() before exit(), to eliminate
      program hang when modem failed to respond
    - ncidd.c: sendLog(): added
    - ncidd.logrotate: changed cidmodem.log to ciddata.log
    - ncid.8: added "SEE ALSO" section, and error code -105
    - ncidd.conf: updated
    - README: updated
    - Makefile: updated
    - ncid.spec: updated
    - TiVo-modem: added
    - INSTALL: updated
    - INSTALL-TiVo: added
    - INSTALL-Win: added
    - ncid: added
    - ncid.1: added
    - changed FORMAT to NCID-FORMAT
    - ncidd.c: main() converted stdin to output to "/dev/null" in
      deamon mode to solve poll problems

Version 0.7
    - ncidd.c split into multiple .c and .h files
    - ncidd.c Ported to BSD and Macintosh, added getopt.c and poll.c
    - ncidd.c formatCID: added MESG stuff
    - ncidd.c formatCID: remove any trailing spaces from NAME field
    - ncidd.c formatCID: modified to output Date, Time, and, Number
      if no NAME field or MESG field found
    - ncidd.c: TIMEOUT 50 -> 100
    - ncid was rewritten and combined with tivocid to replace it
    - added program ncid-speak
    - added program ncid-page
    - added files: ncid.spec, INSTALL-Mac, PROTOCOLS
    - updated files: ncid.1, INSTALL, INSTALL-TiVo, INSTALL-Win,
      NCID-FORMAT, README

Version 0.8
    - ncidd.h and getopt.c: fixed broken getopt
    - Makefile: updated and improved
    - nciddconf.c: fixed bad test for WORDFLAG,
      corrected getAlias() line format comments
    - ncidd.c initModem(): rewrote to eliminate race condition
    - ncidd.c doModem(): changed termios settings so it also
      works with FreeBSD
    - ncidd.conf: added alternate CID settings, and comments on log files
    - ncidd.init: corrected description
    - ncid: fixed test for "--no-gui" to use "=" instead of "=="
    - ncid: added $TCLSH and $WISH variables that are set in a local
      modifications section for FreeBSD since tclsh is named tclVERSION
      and wish is named wishVERSION, where VERSION is the version number
    - ncid: modified to call external programs in background
    - added README-FreeBSD
    - ncid.1: updated
    - ncidd.8: removed ncidd configuration information
    - ncidd.conf.5: added (contains ncidd configuration information)
    - README: updated
    - INSTALL: updated
    - ncid.init: added
    - added the tools subdirectory:
      added cidlog.pl, cidalias.pl, cidlogupd.pl
      added ncidtools.1, README, and Makefile

Version 0.9
    - ncidd.c: formatCID(): patch to get modem data at either character
      position 5 or 7 rathen than position 7 all the time
    - ncidd.c: doALias(): allow "*" in alias[?].from field for
      NMBRDEP and NAMEDEP cases
    - ncidd.h: fixed 1st line of help message (-h)
    - ncidd.conf.5: updated
    - ncid.init: startup is in the background because of a hanging
      problem.  Using "service ncid stop" fails.
    - ncid.spec: updated
    - README: updated
    - CONTRIBUTORS: updated
    - added file input.txt, to test ncid-page and ncid-speak

Version 0.10
    - ncidd.c: modified to send a CIDINFO line indicating the RING count
      added 'send DATA' option and configuration, replaced sendlog option
      and configuration
    - ncidd.h: updated for send
    - nciddconf.c: updated for send
    - ncid: added ring option, updated comments
    - ncid: sendCID(): corrected some comments, pass "\n" at the end
      of the string sent to a external program, use catch with exec
    - ncid: getField(): Fixed error in regex by always searching for
      *FIELDNAME* instead of just FIELDNAME which was not always unique
    - ncid.init: fixed start and stop, added variables
    - Makefile: added passing of MAN variable to tools subdirectory for FreeBSD
    - INSTALL: added a better description of Makefile variables.
    - scripts/ncid-page: changed subject line from "phone number" to
      "Telephone Call" updated wording to indicate cell phone use as
      well as pager
    - scripts/ncid-speak: added Mac OS X interface
    - scripts/ncid-samba: added
    - scripts/README: added
    - test/testclient: server added to test clients using port 3000
    - test/testclient.data: contains data to test a client, used by testclient
    - test/testscript.data: added to test ncid-* scripts
    - test/README: added
    - test/cidtest.log: deleted
    - test/ncidd-test1 renamed to ncid-errors.conf
    - test/ncidd-test2 renamed to ncid-noerrors.conf
    - test/ncidd-test3.conf: deleted
    - test/ncidd-test4.conf: deleted
    - test/ncid-testline: deleted
    - doc/ncidd.8: updated for send cidlog or cidinfo
    - doc/ncidd.conf.5: updated for send cidlog or cidinfo

ncid.sh
#!/bin/sh

# ncid - Network Caller-ID client

# John L. Chmielewski <jlc@cfl.rr.com>
# Copyright (c) 2001, 2002, 2003, 2004, 2005

# ncid is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# any later version.

# ncid is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.

# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA

# START OF LOCAL MODIFICATION SECTION
# set TCLSH variable, FreeBSD will call it something like tclsh8.4 \
TCLSH=tclsh
# set WISH variable, FreeBSD will call it something like wish8.4 \
WISH=wish
# END OF LOCAL MODIFICATION SECTION

# set nice value on TiVo, if "setpri" found \
type setpri > /dev/null 2>&1 && setpri rr 1 $$
# set up TiVo options \
OPTSTIVO="--no-gui --call-prog --program /var/hack/bin/out2osd"
# if name is tivocid, exec tivosh (for backward compatibility) \
case $0 in *tivocid) exec tivosh "$0" $OPTSTIVO "$@"; esac
# look for the --no-gui option \
GUI=""; for i in $*; do if [ "$i" = "--no-gui" ]; then  GUI="$i"; fi; done
# if --no-gui is not specified, look for wish and exec it \
[ -z "$GUI" ] && type $WISH > /dev/null 2>&1 && exec $WISH -f "$0" "$@"
# look for tclsh and exec it \
type $TCLSH > /dev/null 2>&1 && exec $TCLSH "$0" --no-gui "$@"
# look for tivosh and exec it \
type tivosh > /dev/null 2>&1 && exec tivosh "$0" $OPTSTIVO "$@"
# Macintosh \
[ -d /Applications/Wish\ Shell.app ] && \
    /Applications/Wish\ Shell.app/Contents/MacOS/Wish\ Shell -f "$0" "$@"
# tcl or tcl/tk not found \
echo "wish or tclsh or tivosh not found or not in your \$PATH"; exit -1

set ConfigDir   /usr/local/etc/ncid
set ConfigFile  [list $ConfigDir/ncid.conf]

# Constants
set ProgDir     /usr/local/share/ncid
set EXTPROG     ncid-speak

# global variables that can be changed by command line options
set Host        127.0.0.1
set Port        3333
set Delay       60
set Raw         0
set Program     [list $ProgDir/$EXTPROG]
set All         0
set Verbose     0
set NoGUI       0
set Callprog    0
set CallOnRing  0

if {[file exists $ConfigFile]} {
    catch [source $ConfigFile]
}

# global variables that are fixed
set Connect     0
set Count       0
set Ring        0
set Try         0
set Socket      0
set Version     0.61
set VersionInfo "Network CallerID Client Version $Version"
set Usage       {Usage:   ncid  [OPTS] [ARGS]
         OPTS: [--no-gui]
               [--all             | -A]
               [--call-prog       | -C]
               [--delay seconds   | -D seconds]
               [--program PROGRAM | -P PROGRAM]
               [--raw             | -R]
               [--ring count      | -r count]
               [--verbose         | -V]
         ARGS: [IP_ADDRESS        | HOSTNAME]
               [PORT_NUMBER]}

set About \
"
Network Caller ID Client: NCID Version $Version
Copyright (C) 2001, 2002, 2003, 2004, 2005
by John L. Chmielewski
"

# display error message and exit
proc exitMsg {code msg} {
    global NoGUI

    if $NoGUI {
        puts stderr $msg
    } else {
        option add *Dialog.msg.wrapLength 9i
        option add *Dialog.msg.font "courier 12"
        tk_messageBox -message $msg -type ok
    }
    exit $code
}

# display TCP/IP error message, then try to reconnect
proc errorMsg {msg} {
    global Connect
    global Count
    global Delay
    global Try
    global Txt
    global NoGUI
    global Verbose

    if $NoGUI {
        if $Verbose {puts -nonewline stderr $msg}
        after [expr $Delay*1000] retryConnect
    } else {
        set Connect 0
        set Count $Delay
        while {$Count > 0} {
            if {$Count == 1} {
                set Txt "$msg Try $Try in $Count second."
            } else {
                set Txt "$msg Try $Try in $Count seconds."
            }
            set Count [expr $Count - 1]
            after [expr 1000] waitOnce
            vwait Once
        }
        retryConnect
    }
}

# Count loop delay
proc waitOnce {} {
    global Once

    set Once 0
    after cancel waitOnce
}

# try to connect to CID server again
proc retryConnect {} {
    global Host
    global Port
    global NoGUI

    if $NoGUI {after cancel retryConnect}
    connectCID $Host $Port
}

# This catches a lot of errors!
proc bgerror {mess} {
    global errorInfo

    exitMsg 1 "BGError: $mess\n"
}

# Get data from CID server
proc getCID {} {
    global Connect
    global Socket
    global Host
    global Port
    global Raw
    global Try
    global VersionInfo
    global NoGUI
    global Callprog
    global Ring
    global CallOnRing
    global cid

    set msg {CID connection closed}
    set cnt 0
    while {$cnt != -1} {
        if {[eof $Socket] || [catch {set cnt [gets $Socket dataBlock]} msg]} {
            # remove event handler
            fileevent $Socket readable ""
            close $Socket
            set Try [expr $Try + 1]
            errorMsg "$Host:$Port - $msg\n"
            return
        }
        set Try 0
        # remove <CR> should be \r$ instead of .$, but doesn't work on TiVo
        regsub {(.*).$} $dataBlock {\1} dataBlock

        if $Raw {
            # output NCID data line
            displayLog [list $dataBlock]
            if {[string match 200* $dataBlock]} {
                if {!$NoGUI} { displayCID "$VersionInfo\n$dataBlock" }
            }
        } else {
            if {[string match 200* $dataBlock]} {
                # output NCID server connect message
                regsub {200 (.*)} $dataBlock {\1} dataBlock
                if $NoGUI { displayLog [list "$VersionInfo\n$dataBlock"]
                } else { displayCID "$VersionInfo\n$dataBlock" }
            }
        }
        if {[set type [checkCID $dataBlock]]} {
            if {$CallOnRing && $type == 5 && $Callprog} {
                # found CIDINFO line
                set cidinfo [getField RING $dataBlock]
                if {$Ring == $cidinfo} {sendCID $cid}
            } elseif {$type == 4} {
                # found MSG line
                regsub {MSG: (.*)} $dataBlock {\1} msg
                displayLog [list $msg]
            } else {
                # found CID, EXTRA, or CIDLOG line
                set cid [formatCID $dataBlock]
                # Clear display window if CIDLOG line is received
                if {!$Connect && $type == 3} {
                    set Connect 1
                    if {!$NoGUI} {clearLog}
                }
                # display CID log
                if {$type < 4} {
                    if {!$Raw} {displayLog $cid}
                }
                # display CID
                if {$type < 3} {
                    if {!$NoGUI} {
                        displayCID $cid
                        # create a pop-up, even if open, in current desktop
                        wm iconify .
                        update
                        wm deiconify .
                        raise .
                        update
                    }
                    if {!$CallOnRing && $Callprog} {sendCID $cid}
                }
            }
        }
    }
}

proc checkCID {dataBlock} {
    # ignore all but the CID or EXTRA information line
    if [string match CID:* $dataBlock] {return 1}
    if [string match *EXTRA:* $dataBlock] {return 2}
    if [string match CIDLOG:* $dataBlock] {return 3}
    if [string match MSG:* $dataBlock] {return 4}
    if [string match CIDINFO:* $dataBlock] {return 5}
    return 0
}

# must be sure the line passed checkCID
proc formatCID {dataBlock} {
    set cidname [getField NAME $dataBlock]
    set cidnumber [getField NU*MBE*R $dataBlock]
    if {![regsub \
        {1?([0-9][0-9][0-9])([0-9][0-9][0-9])([0-9][0-9][0-9][0-9])} \
        $cidnumber {(\1)\2-\3} cidnumber]} {
        regsub {([0-9][0-9][0-9])([0-9][0-9][0-9][0-9])} \
        $cidnumber {\1-\2} cidnumber
    }
    set ciddate [getField DATE $dataBlock]
    if {![regsub {([0-9][0-9])([0-9][0-9])([0-9][0-9][0-9][0-9])} \
        $ciddate {\1/\2/\3} ciddate]} {
        regsub {([0-9][0-9])([0-9][0-9].*)} $ciddate {\1/\2} ciddate
    }
    set cidtime [getField TIME $dataBlock]
    regsub {([0-9][0-9])([0-9][0-9])} $cidtime {\1:\2} cidtime
    if [string match {*\*LINE\**} $dataBlock] {
        set cidline [getField LINE $dataBlock]
        regsub {(.*)} $cidline {<\1>} cidline
    } else {set cidline <>}
    regsub {<-*>} $cidline {} cidline

    return [list $ciddate $cidtime $cidnumber $cidname $cidline]
}

# get a field from the CID data
proc getField {dataString dataBlock} {
    regsub ".*\\*$dataString\\*" $dataBlock {} result
    regsub {([\w\s]*)\*.*} $result {\1} result
    return $result
}

# pass the CID information to an external program
# Output: "$ciddate\n$cidtime\n$cidnumber\n$cidname\n"
proc sendCID {cid} {
    global All
    global Program

    if $All {
        # pass DATE\nTIME\nNUMBER\nNAME\n
        catch {exec [lindex $Program 0] << \
            "[lindex $cid 0]\n[lindex $cid 1]\n[lindex $cid 2]\n[lindex $cid 3]\n" &} oops
        # pass NAME NUMBER
        } else {
            catch {exec [lindex $Program 0] << \
                "[lindex $cid 3] [lindex $cid 2]\n" &} oops
        }
}

# display CID information
# Input: "$ciddate $cidtime $cidnumber $cidname $cidline\n"
proc displayCID {cid} {
    global Txt

    if {[llength $cid] != 5} { set Txt $cid
    } else { set Txt "[lindex $cid 3]\n[lindex $cid 2]" }
    update
}

# display Call Log
# Input: "$ciddate $cidtime $cidnumber $cidname $cidline\n"
proc displayLog {cid} {
    global Verbose
    global NoGUI
    if $NoGUI {
        if $Verbose {
            puts "[lindex $cid 0] [lindex $cid 1] [lindex $cid 4] [lindex $cid 2] [lindex $cid 3]"
        }
    } else {
        set cid "[lindex $cid 0] [lindex $cid 1] [lindex $cid 4] [lindex $cid 2] [lindex $cid 3]\n"
        .vl configure -state normal
        .vl insert end $cid
        .vl yview moveto 1.0
        .vl configure -state disabled

    }
}

# Open a connection to the CID server
proc connectCID {Host Port} {
    global Try
    global Socket
    global NoGUI

    # open socket to server
    if {[catch {set Socket [socket -async $Host $Port]} msg]} {
        set Try [expr $Try + 1]
        errorMsg "$Host:$Port - $msg\n"
    } else {
        # set socket to binary I/O and non-blocking
        fconfigure $Socket -translation {binary binary} -blocking 0 
        # get response from server as event
        fileevent $Socket readable getCID
        if $NoGUI { displayLog "Connecting to $Host:$Port"
        } else { displayCID "Connecting to\n$Host:$Port"}
    }
}

proc getArg {} {
    global argc
    global argv
    global All
    global Raw
    global Host
    global Port
    global Delay
    global Usage
    global NoGUI
    global Verbose
    global Program
    global Callprog
    global Ring
    global CallOnRing
    global ProgDir

    for {set cnt 0} {$cnt < $argc} {incr cnt} {
        set optarg [lindex $argv [expr $cnt + 1]]
        switch -regexp -- [set opt [lindex $argv $cnt]] {
            {^-r$} -
            {^--ring$} {
                incr cnt
                if {$optarg != ""
                    && [regexp {^[023456789][0-9]?$} $optarg]} {
                    set Ring $optarg
                    set CallOnRing 1
                } else {exitMsg 4 "Invalid $opt argument: $optarg\n$Usage\n"}
            }
            {^-N$} -
            {^--no-gui$} {set NoGUI 1}
            {^-A$} -
            {^--all$} {set All 1}
            {^-C$} -
            {^--call-prog$} {set Callprog 1}
            {^-D$} -
            {^--delay$} {
                incr cnt
                if {$optarg != ""
                    && [regexp {^[0-9]+$} $optarg]} {
                    set Delay $optarg
                } else {exitMsg 4 "Invalid $opt argument: $optarg\n$Usage\n"}
            }
            {^-P$} -
            {^--program$} {
                incr cnt
                if {$optarg != ""
                    && [regexp {^[./\0]+} $optarg]} {
                    set Program [list $optarg]
                } else {set Program [list $ProgDir/$optarg]}
            }
            {^-V$} -
            {^--verbose$} {set Verbose 1}
            {^-R$} -
            {^--raw$} {set Raw 1}
            {^[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+$} {set Host $opt}
            {^[A-Za-z]+[.A-Za-z0-9-]+$} {set Host $opt}
            {^[0-9]+$} {set Port $opt}
            default {exitMsg 5 "Unknown option: $opt\n$Usage\n"}
        }
    }
}

proc makeWindow {} {
    frame .fr -borderwidth 2
    wm title . "Network Caller ID"
    wm protocol . WM_DELETE_WINDOW exit
    wm resizable . 0 0
    pack .fr

    frame .menubar -relief raised -bd 2
    pack .menubar -in .fr -fill x

    label .vc -textvariable Txt -font {Helvetica -14 bold}
    pack .vc

    text .vl -width 62 -height 5 -yscrollcommand ".ys set" \
        -state disabled -font {Courier -14 bold}
    scrollbar .ys -command ".vl yview"
    pack .vl .ys -in .fr -side left -fill y

    menubutton .menubar.file -text File -underline 0 -menu .menubar.file.menu
    pack .menubar.file -side left
    menubutton .menubar.help -text Help -underline 0 -menu .menubar.help.menu
    pack .menubar.help -side right

    menu .menubar.file.menu -tearoff 0
    .menubar.file.menu add command -label "Clear Log" -command clearLog
    .menubar.file.menu add command -label "Reconnect" -command reconnect
    .menubar.file.menu add command -label Quit -command exit

    menu .menubar.help.menu -tearoff 0
    .menubar.help.menu add command -label About -command aboutPopup
}

proc aboutPopup {} {
    global About

    option add *Dialog.msg.wrapLength 9i
    option add *Dialog.msg.font "Helvetica 12"
    tk_messageBox -message $About -type ok -title About
}

proc clearLog {} {

    .vl configure -state normal
    .vl delete 1.0 end
    .vl yview moveto 0.0
    .vl configure -state disabled
}

proc reconnect {} {
    global Connect
    global Socket
    global Once
    global Count
    global Host
    global Port

    # connection to server already closed
    if $Count {
        set Count 0
        return
    }

    # close connection to server
    flush $Socket
    fileevent $Socket readable ""
    close $Socket

    # delay for 0.1 seconds
    after [expr 100] waitOnce
    vwait Once

    set Connect 0
    retryConnect
}

getArg
if {!$NoGUI} makeWindow
if $Callprog {
    if {[file exists $Program]} {
        if {![file executable $Program]} {
            exitMsg 2 "Program Not Executable: $Program"
        }
    } else { exitMsg 3 "Program Not Found: $Program" }
} else { set Verbose 1 }
connectCID $Host $Port

# enter event loop
vwait forever
